### Importing all the necessary libraries

In [1]:
import numpy as np
from textblob import TextBlob
import pandas as pd
from nltk.corpus import stopwords
import string
import re
import nltk
import pandas as pd
nltk.download('wordnet')
from nltk.stem import WordNetLemmatizer
from textblob import  TextBlob


pd.set_option('display.max_colwidth' , 330)

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\tsamr\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [2]:
source = 'https://sentimentanalysisdata.blob.core.windows.net/scraped-data/amazon.csv?sp=r&st=2022-11-19T23:09:56Z&se=2022-11-20T07:09:56Z&spr=https&sv=2021-06-08&sr=b&sig=M6kgyxokrG5LFgZea1lLI0CH7jHk52jn39cXKx0%2Bf6M%3D'

In [3]:
df = pd.read_csv(source)
df

StarRating  \
0              1   
1              3   
2              1   
3              3   
4              1   
...          ...   
2495           1   
2496           1   
2497           1   
2498           3   
2499           3   

                                                                                                                                                                                                                                                                                                                                         Review  
0     They just updated the app so that when you go to shop and you look up a product it autoplays a video on every. single. product. And when you go to click on the product that you may be interested in, it starts blasting the product video. I was advised there is no way to turn this off unless I also wanted to stop autoplay on m...  
1     I wish the app would work properly, nothing seems to want to load. It keeps freezing and clicking to things when I haven't touched my phone(as I left it alone to let it load). Also keeps exiting me out, hasn't completely crashed, but definitely likes to take me back to home. Nov 16, even after update, still having the same i...  
2     It's difficult to find what I'm looking for. There are so many unrelated results, the filters are usually for things I don't care about, and many vendors (especially China-based ones) list innacurate or misleading information. Account settings are a nightnare to navigate; there's not even an option to search the settings (en...  
3     It's hard to find exactly what you're looking for and sometimes it excludes results for some reason. Heaven forbid you try and move something from one list to another or delete something off a list. Everytime I try the app immediately stalls and quits working within a few minutes. I never get more than 7 things deleted at a ...  
4     Horrible algorithm, you have to know exactly how to search to find what you want, and your home page will be filled with the most useless items for you. It crashes, loads slow, bugs, and is generally inconvenient. Also they keep updating what the app looks like and every time it gets uglier and more inconvenient. The orders ...  
...                                                                                                                                                                                                                                                                                                                                         ...  
2495                                                                                                                                                                      Getting worse. Now 'watch lists' no longer show used prices. You have to click each individual item to bring up lowest prices for anything used. Website still works.  
2496  I created an account and ordered it you guys sent me an email confirming my password was ok. Now when I put in my telephone number it says it belongs to another.Also when I go to log out it does not log me out. That means anyone can log in because it remains open your app is difficult to navigate and it's he frustrating just...  
2497  Delivery drivers ignore delivery instructions. No way to contact amazon. Then they keep your momey for 5 extra days to "ensure delivery" before they give you a refund. Ive had packages left in the rain, my mailbox (notified postmaster general), on the bumper of my truck (didnt see that one until i ran over it). Granted fedex...  
2498                                                                                                                                                                                                                                        Great app, but missing the all important dark mode. It is 2022, no app should be missing dark mode!  
2499                                           

In [4]:
df.sample(10)

StarRating  \
688            5   
820            4   
998            5   
185            1   
772            1   
15             1   
440            1   
1980           2   
279            4   
1773           3   

                                                                                                                                                                                                                                                                                                                                         Review  
688                                                                                                               A fantastic service! I love the ease and haste of the service. I also love the notifications and emails giving me updates every step of the way. People are crazy for hating this company, yet they still use it. It's great!  
820                                                                                                                                                          Everything is good but membership is a little high. The app UI is nice but one thing I don't like is the unrelated videos of the product when you scroll down on the product page.  
998   I love shopping and Amazon definitely makes it much easier. 98.5% of the time I find exactly what I'm looking for and shipping is also timely. I really appreciate that they have so many tools for you to ensure that the item you want is really what your looking for. My favorite part are the reviews, where buyers can express w...  
185                                                                                                                 Not a bad app but everytime someone sends me a link it switches to their language and I have to go to the the settings and change it back. It is quite annoying. The app is now sending adds in the notifications= one star  
772   Since 10/22, Map notifications are sent after delivery now - when app is opened. USELESS 9/22 Why are shipment tracking notifications now only sent after I open the app instead of hours earlier like before? 6/22 After recently re-doing the app, why is there still no direct link to the Orders page like the website? 2022: Need...  
15    I love Amazon but this app needs A LOT of work. It freezes every few seconds the ENTIRE time I try using it. It makes me not even want to shop because it take hours to get anything to load and especially when you are trying to read reviews and look at customers pictures with their review. It is impossible to shop around with...  
440   The app is underperforming. Most of the tabs are not loading data but the cute puppies saying something went wrong... I am using android device and the app has no updates to be actioned so I am not sure.. but this behaviour happened in the past 1-2 months and I am very very very dissapointment. I can't use the app .. rather ...  
1980                                                                                                                                     Poor interface. When you're browsing the app and you hit 'view as website/desktop' once, there is no way to make it go back to an app interface which makes scrolling and searching much, much harder.  
279        Works well enough, but would be better with less pictures and adverts clogging up the screen when using it to try and see information about orders already placed...... its good for browsing/shopping, but a little bit laborious finding information about orders already 'in progress'. No major issues though - so still 4 stars  
1773  Works ok, better than the website on mobile, but I sometimes have trouble with wishlists refusing to load properly, and in recent months when the app tries to load the map for a delivery nearby it just fails and removes the map. As tracking my deliveries is a major reason for using the app, this is extremely annoying. It doe...

In [5]:
#lowercase all the reviews data
df['Review'] = df['Review'].str.lower()
df.head()

StarRating  \
0           1   
1           3   
2           1   
3           3   
4           1   

                                                                                                                                                                                                                                                                                                                                      Review  
0  they just updated the app so that when you go to shop and you look up a product it autoplays a video on every. single. product. and when you go to click on the product that you may be interested in, it starts blasting the product video. i was advised there is no way to turn this off unless i also wanted to stop autoplay on m...  
1  i wish the app would work properly, nothing seems to want to load. it keeps freezing and clicking to things when i haven't touched my phone(as i left it alone to let it load). also keeps exiting me out, hasn't completely crashed, but definitely likes to take me back to home. nov 16, even after update, still having the same i...  
2  it's difficult to find what i'm looking for. there are so many unrelated results, the filters are usually for things i don't care about, and many vendors (especially china-based ones) list innacurate or misleading information. account settings are a nightnare to navigate; there's not even an option to search the settings (en...  
3  it's hard to find exactly what you're looking for and sometimes it excludes results for some reason. heaven forbid you try and move something from one list to another or delete something off a list. everytime i try the app immediately stalls and quits working within a few minutes. i never get more than 7 things deleted at a ...  
4  horrible algorithm, you have to know exactly how to search to find what you want, and your home page will be filled with the most useless items for you. it crashes, loads slow, bugs, and is generally inconvenient. also they keep updating what the app looks like and every time it gets uglier and more inconvenient. the orders ...

In [6]:
contractions = { 
"ain't": "am not",
"aren't": "are not",
"can't": "cannot",
"can't've": "cannot have",
"'cause": "because",
"could've": "could have",
"couldn't": "could not",
"couldn't've": "could not have",
"didn't": "did not",
"doesn't": "does not",
"don't": "do not",
"hadn't": "had not",
"hadn't've": "had not have",
"hasn't": "has not",
"haven't": "have not",
"he'd": "he had",
"he'd've": "he would have",
"he'll": "he will",
"he'll've": "he shall have ",
"he's": "e is",
"how'd": "how did",
"how'd'y": "how do you",
"how'll": "how will",
"how's": "how has",
"i'd": "i would",
"i'd've": "i would have",
"i'll": "i will",
"i'll've": "i will have",
"i'm": "i am",
"i've": "i have",
"isn't": "is not",
"it'd": "it would",
"it'd've": "it would have",
"it'll": "it will",
"it'll've": "it will have",
"it's": "it is",
"let's": "let us",
"ma'am": "madam",
"mayn't": "may not",
"might've": "might have",
"mightn't": "might not",
"mightn't've": "might not have",
"must've": "must have",
"mustn't": "must not",
"mustn't've": "must not have",
"needn't": "need not",
"needn't've": "need not have",
"o'clock": "of the clock",
"oughtn't": "ought not",
"oughtn't've": "ought not have",
"shan't": "shall not",
"sha'n't": "shall not",
"shan't've": "shall not have",
"she'd": "she had",
"she'd've": "she would have",
"she'll": "she will",
"she'll've": "she will have",
"she's": "she is",
"should've": "should have",
"shouldn't": "should not",
"shouldn't've": "should not have",
"so've": "so have",
"so's": "so is",
"that'd": "that would",
"that'd've": "that would have",
"that's": "that has",
"there'd": "there would",
"there'd've": "there would have",
"there's": "there is",
"they'd": "they would",
"they'd've": "they would have",
"they'll": "they will",
"they'll've": "they will have",
"they're": "they are",
"they've": "they have",
"to've": "to have",
"wasn't": "was not",
"we'd": "we would",
"we'd've": "we would have",
"we'll": "we will",
"we'll've": "we will have",
"we're": "we are",
"we've": "we have",
"weren't": "were not",
"what'll": "what will",
"what'll've": "what will have",
"what're": "what are",
"what's": "what is",
"what've": "what have",
"when's": "when is",
"when've": "when have",
"where'd": "where did",
"where's": "where is",
"where've": "where have",
"who'll": "who will",
"who'll've": "who will have",
"who's": "who is",
"who've": "who have",
"why's": "why is",
"why've": "why have",
"will've": "will have",
"won't": "will not",
"won't've": "will not have",
"would've": "would have",
"wouldn't": "would not",
"wouldn't've": "would not have",
"y'all": "you all",
"y'all'd": "you all would",
"y'all'd've": "you all would have",
"y'all're": "you all are",
"y'all've": "you all have",
"you'd": "you would",
"you'd've": "you would have",
"you'll": "you will",
"you'll've": "you will have",
"you're": "you are",
"you've": "you have"
}


In [7]:
contractions_re=re.compile('(%s)' % '|'.join(contractions.keys()))
def expand_contractions(text,contractions=contractions):
    def replace(match):
        return contractions[match.group(0)]
    return contractions_re.sub(replace, text)
# Expanding Contractions in the reviews
df['Review']=df['Review'].apply(lambda x:expand_contractions(x))
df.head()

StarRating  \
0           1   
1           3   
2           1   
3           3   
4           1   

                                                                                                                                                                                                                                                                                                                                      Review  
0  they just updated the app so that when you go to shop and you look up a product it autoplays a video on every. single. product. and when you go to click on the product that you may be interested in, it starts blasting the product video. i was advised there is no way to turn this off unless i also wanted to stop autoplay on m...  
1  i wish the app would work properly, nothing seems to want to load. it keeps freezing and clicking to things when i have not touched my phone(as i left it alone to let it load). also keeps exiting me out, has not completely crashed, but definitely likes to take me back to home. nov 16, even after update, still having the same...  
2  it is difficult to find what i am looking for. there are so many unrelated results, the filters are usually for things i do not care about, and many vendors (especially china-based ones) list innacurate or misleading information. account settings are a nightnare to navigate; there is not even an option to search the settings...  
3  it is hard to find exactly what you are looking for and sometimes it excludes results for some reason. heaven forbid you try and move something from one list to another or delete something off a list. everytime i try the app immediately stalls and quits working within a few minutes. i never get more than 7 things deleted at ...  
4  horrible algorithm, you have to know exactly how to search to find what you want, and your home page will be filled with the most useless items for you. it crashes, loads slow, bugs, and is generally inconvenient. also they keep updating what the app looks like and every time it gets uglier and more inconvenient. the orders ...

In [8]:
# removing punctuation

print(string.punctuation)

def remove_punc(txt):
    txt_nopunct = "".join([x for x in txt if x not in string.punctuation])
    return txt_nopunct

df['Review'] = df['Review'].apply(lambda x: remove_punc(x))
df.head()

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


StarRating  \
0           1   
1           3   
2           1   
3           3   
4           1   

                                                                                                                                                                                                                                                                                                                                      Review  
0  they just updated the app so that when you go to shop and you look up a product it autoplays a video on every single product and when you go to click on the product that you may be interested in it starts blasting the product video i was advised there is no way to turn this off unless i also wanted to stop autoplay on my ama...  
1  i wish the app would work properly nothing seems to want to load it keeps freezing and clicking to things when i have not touched my phoneas i left it alone to let it load also keeps exiting me out has not completely crashed but definitely likes to take me back to home nov 16 even after update still having the same issues i ...  
2  it is difficult to find what i am looking for there are so many unrelated results the filters are usually for things i do not care about and many vendors especially chinabased ones list innacurate or misleading information account settings are a nightnare to navigate there is not even an option to search the settings enterin...  
3  it is hard to find exactly what you are looking for and sometimes it excludes results for some reason heaven forbid you try and move something from one list to another or delete something off a list everytime i try the app immediately stalls and quits working within a few minutes i never get more than 7 things deleted at a t...  
4  horrible algorithm you have to know exactly how to search to find what you want and your home page will be filled with the most useless items for you it crashes loads slow bugs and is generally inconvenient also they keep updating what the app looks like and every time it gets uglier and more inconvenient the orders page has...

In [9]:
#removing stopwords
sr = stopwords.words('english')

In [10]:
df['Review'] = df['Review'].apply(lambda x: ' '.join(x for x in x.split() if x not in sr))
df.head()

StarRating  \
0           1   
1           3   
2           1   
3           3   
4           1   

                                                                                                                                                                                                                                                                                                                                      Review  
0                                                 updated app go shop look product autoplays video every single product go click product may interested starts blasting product video advised way turn unless also wanted stop autoplay amazon prime videos means watching show would autoplay next episode fail understand two would linked  
1                    wish app would work properly nothing seems want load keeps freezing clicking things touched phoneas left alone let load also keeps exiting completely crashed definitely likes take back home nov 16 even update still issues close app reopen get work everytime click something want look buy cart images load blurry  
2  difficult find looking many unrelated results filters usually things care many vendors especially chinabased ones list innacurate misleading information account settings nightnare navigate even option search settings entering search terms search bar settings screen returns product results seem way choose preferred communicat...  
3   hard find exactly looking sometimes excludes results reason heaven forbid try move something one list another delete something list everytime try app immediately stalls quits working within minutes never get 7 things deleted time good thing never issues orders arrive within projected time website works well app extremely buggy  
4                                         horrible algorithm know exactly search find want home page filled useless items crashes loads slow bugs generally inconvenient also keep updating app looks like every time gets uglier inconvenient orders page changed past orders current combined mixed making orders extremely hard visualize

In [11]:
#removing numbers

df['Review'] = df['Review'].apply(lambda x: ' '.join(x for x in x.split() if not x.isdigit()))
df.head(10)


StarRating  \
0           1   
1           3   
2           1   
3           3   
4           1   
5           1   
6           2   
7           2   
8           1   
9           1   

                                                                                                                                                                                                                                                                                                                                      Review  
0                                                 updated app go shop look product autoplays video every single product go click product may interested starts blasting product video advised way turn unless also wanted stop autoplay amazon prime videos means watching show would autoplay next episode fail understand two would linked  
1                       wish app would work properly nothing seems want load keeps freezing clicking things touched phoneas left alone let load also keeps exiting completely crashed definitely likes take back home nov even update still issues close app reopen get work everytime click something want look buy cart images load blurry  
2  difficult find looking many unrelated results filters usually things care many vendors especially chinabased ones list innacurate misleading information account settings nightnare navigate even option search settings entering search terms search bar settings screen returns product results seem way choose preferred communicat...  
3     hard find exactly looking sometimes excludes results reason heaven forbid try move something one list another delete something list everytime try app immediately stalls quits working within minutes never get things deleted time good thing never issues orders arrive within projected time website works well app extremely buggy  
4                                         horrible algorithm know exactly search find want home page filled useless items crashes loads slow bugs generally inconvenient also keep updating app looks like every time gets uglier inconvenient orders page changed past orders current combined mixed making orders extremely hard visualize  
5        stale app sadly support dark mode search poor showing many irrelevant results order history shortened months complete history full history scroll helpful finding items reorder see items purchased warranty filter year cannot remember year much difficult search less useful handling missing orders sucks november improvements  
6                                                                                                             wish list feature almost unusable months near impossible scroll list constantly snapping back top editing comments priorities actually impossible almost always ends app crashing store entirely web based pretty embarrassing  
7                                        useful broken orders screen load majority time gets stuck frozen makes reordering items handling returns checking order status sharing product links much difficult broken months reported bug previously response place many orders people probably may problem fyi workaround go buy search order  
8  since new update asked repeatedly small time frame verify bot completely glitches even verify viewingtapping anything n editing wishlists leads something went wrong end message next minutes app impossible use constant error messages saving wishlist glitchy misclicks clicks items even screen slow response tapping add list uni...  
9                                                                                                        horrible algo filters results unrelated filters even work apply filter price range shows products higher lower filters sucks makes hard navigate look desired product understand amazon cannot work feature make smooth useful user

In [12]:
#lemmatization
lemmatizer = WordNetLemmatizer()

def lemmatize_words(text):
    return " ".join([lemmatizer.lemmatize(word) for word in text.split()])


df["Review"] = df["Review"].apply(lambda x: lemmatize_words(x))
df.head()

StarRating  \
0           1   
1           3   
2           1   
3           3   
4           1   

                                                                                                                                                                                                                                                                                                                                  Review  
0                                                updated app go shop look product autoplays video every single product go click product may interested start blasting product video advised way turn unless also wanted stop autoplay amazon prime video mean watching show would autoplay next episode fail understand two would linked  
1                         wish app would work properly nothing seems want load keep freezing clicking thing touched phoneas left alone let load also keep exiting completely crashed definitely like take back home nov even update still issue close app reopen get work everytime click something want look buy cart image load blurry  
2  difficult find looking many unrelated result filter usually thing care many vendor especially chinabased one list innacurate misleading information account setting nightnare navigate even option search setting entering search term search bar setting screen return product result seem way choose preferred communication method  
3        hard find exactly looking sometimes excludes result reason heaven forbid try move something one list another delete something list everytime try app immediately stall quits working within minute never get thing deleted time good thing never issue order arrive within projected time website work well app extremely buggy  
4                                               horrible algorithm know exactly search find want home page filled useless item crash load slow bug generally inconvenient also keep updating app look like every time get uglier inconvenient order page changed past order current combined mixed making order extremely hard visualize

Now, our data is clean. We have removed all the stopwords, punctuation, numbers, contractions, lower cased every text, lemmatized them. The data is read for sentimental analysis. For this we will be using the TextBlob library. 

In [13]:
sentiment=[]
sentiment_score = []
keywords = []

for x in df['Review'] :
    senti=TextBlob(x) ## sentiment analysis using TextBlob
    
    Token= [t for t in x.split()] ## string the tokens/keywords
    keywords.append(Token)

    ## conditions for division of sentiment
    if senti.sentiment.polarity > 0 and senti.sentiment.polarity <= 1 :
        sentiment_score.append(round(senti.sentiment.polarity,3))
        sentiment.append("Postive")
    elif senti.sentiment.polarity == 0 :
        sentiment_score.append(round(senti.sentiment.polarity,3))
        sentiment.append("Neutral")
    else:
        sentiment_score.append(round(senti.sentiment.polarity,3))
        sentiment.append("Negative")

In [14]:
## storing the new data in our dataset
df['Keywords']=keywords
df['Sentiment']=sentiment
df['Sentiment_Score'] = sentiment_score

In [15]:
df.head(10)

StarRating  \
0           1   
1           3   
2           1   
3           3   
4           1   
5           1   
6           2   
7           2   
8           1   
9           1   

                                                                                                                                                                                                                                                                                                                                      Review  \
0                                                    updated app go shop look product autoplays video every single product go click product may interested start blasting product video advised way turn unless also wanted stop autoplay amazon prime video mean watching show would autoplay next episode fail understand two would linked   
1                             wish app would work properly nothing seems want load keep freezing clicking thing touched phoneas left alone let load also keep exiting completely crashed definitely like take back home nov even update still issue close app reopen get work everytime click something want look buy cart image load blurry   
2      difficult find looking many unrelated result filter usually thing care many vendor especially chinabased one list innacurate misleading information account setting nightnare navigate even option search setting entering search term search bar setting screen return product result seem way choose preferred communication method   
3            hard find exactly looking sometimes excludes result reason heaven forbid try move something one list another delete something list everytime try app immediately stall quits working within minute never get thing deleted time good thing never issue order arrive within projected time website work well app extremely buggy   
4                                                   horrible algorithm know exactly search find want home page filled useless item crash load slow bug generally inconvenient also keep updating app look like every time get uglier inconvenient order page changed past order current combined mixed making order extremely hard visualize   
5                 stale app sadly support dark mode search poor showing many irrelevant result order history shortened month complete history full history scroll helpful finding item reorder see item purchased warranty filter year cannot remember year much difficult search le useful handling missing order suck november improvement   
6                                                                                                                  wish list feature almost unusable month near impossible scroll list constantly snapping back top editing comment priority actually impossible almost always end app crashing store entirely web based pretty embarrassing   
7                                                useful broken order screen load majority time get stuck frozen make reordering item handling return checking order status sharing product link much difficult broken month reported bug previously response place many order people probably may problem fyi workaround go buy search order   
8  since new update asked repeatedly small time frame verify bot completely glitch even verify viewingtapping anything n editing wishlists lead something went wrong end message next minute app impossible use constant error message saving wishlist glitchy misclicks click item even screen slow response tapping add list uninstalli...   
9                                                                                                                horrible algo filter result unrelated filter even work apply filter price range show product higher lower filter suck make hard navigate look desired product understand amazon cannot work feature make smooth useful user   

                                                                                                                      

In [16]:
df['Sentiment'].value_counts()

Postive     1239
Negative    1055
Neutral      206
Name: Sentiment, dtype: int64

In [17]:
### saving the dataset 
df.to_csv("amazon_sentiment_analysis.csv",index= False)

So, after the sentimental analysis with the TextBlob library, we got 1239 positive values, 1055 negative value and 206 neural values which is a good collection of data.